tests/ui/new_ret_no_self.rs

```diff
@@ -91,3 +91,31 @@ impl V {
         unimplemented!();
     }
 }
+
+struct TupleReturnerOk;
+
+impl TupleReturnerOk {
+    // should not trigger lint
+    pub fn new() -> (Self, u32) { unimplemented!(); }
```

Comment:
`((Self, u32), u32)` Will also trigger the lint. I'm not sure at all how strict we should be. I mean it seems perfectly reasonable to allow `(Option<Self>, u32)` or at least `Option<(Self, u32)>`. This might get out of hand quickly.

If you want to cover a lot of cases, you can call [`walk`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyS.html#method.walk) on the return type (giving you an iterator) and iterate over everything.

That would still cause false positives for `struct Foo(pub Bar); impl Bar { fn new() -> Foo { ... } }` but we need to restrict ourselves somewhere I guess.